src/kernel/page.c

```diff
@@ -0,0 +1,531 @@
+
+    page_init_debug("\nmap_level: table_ptr ");
+    page_init_debug_u64(u64_from_pointer(table_ptr));
+    page_init_debug(", level ");
+    page_init_debug_u64(level);
+    page_init_debug("\n   v ");
+    page_init_debug_u64(v.start);
+    page_init_debug(" - ");
+    page_init_debug_u64(v.end);
+    page_init_debug(", p ");
+    page_init_debug_u64(*p);
+    page_init_debug(" first ");
+    page_init_debug_u64(first_index);
+    page_init_debug(" last ");
+    page_init_debug_u64(last_index);
+    page_init_debug("\n");
+    assert(first_index <= last_index);
+    assert(table_ptr && table_ptr != INVALID_ADDRESS);
+
+    for (int i = first_index; i <= last_index; i++, next_addr(v.start, mask)) {
+        boolean invalidate = false; /* page at v.start */
+        page_init_debug("   index ");
+        page_init_debug_u64(i);
+        page_init_debug(", v.start ");
+        page_init_debug_u64(v.start);
+        page_init_debug(", p ");
+        page_init_debug_u64(*p);
+        u64 pte = table_ptr[i];
+        page_init_debug(", pte ");
+        page_init_debug_u64(pte);
+        page_init_debug("\n");
+        if (!pte_is_present(pte)) {
+            if (level == PT_PTE_LEVEL) {
+                pte = page_pte(*p, flags);
+                next_addr(*p, mask);
+                invalidate = true;
+            } else if (!flags_has_minpage(flags) && level > PT_FIRST_LEVEL &&
+                       (pagemem.levelmask & U64_FROM_BIT(level)) && (v.start & mask) == 0 &&
+                       (*p & mask) == 0 && range_span(v) >= U64_FROM_BIT(shift)) {
+                pte = block_pte(*p, flags);
+                next_addr(*p, mask);
+                invalidate = true;
+            } else {
+                page_init_debug("      new level: ");
+                void *tp;
+                u64 tp_phys;
+                if ((tp = allocate_table_page(&tp_phys)) == INVALID_ADDRESS) {
+                    msg_err("failed to allocate page table memory\n");
+                    return false;
+                }
+                /* user and writable are AND of flags from all levels */
+                pte = new_level_pte(tp_phys);
+                u64 end = vlbase | (((u64)(i + 1)) << shift);
+                /* length instead of end to avoid overflow at end of space */
+                u64 len = MIN(range_span(v), end - v.start);
+                page_init_debug("  end ");
+                page_init_debug_u64(end);
+                page_init_debug(", len ");
+                page_init_debug_u64(len);
+                page_init_debug("\n");
+                if (!map_level(tp, level + 1, irangel(v.start, len), p, flags, fe))
+                    return false;
+            }
+            page_init_debug("      pte @ ");
+            page_init_debug_u64(u64_from_pointer(&table_ptr[i]));
+            page_init_debug(" = ");
+            page_init_debug_u64(pte);
+            page_init_debug("\n");
+            table_ptr[i] = pte;
+            if (invalidate)
```

Comment:
We might consider calling an invalidate on this address here or set invalidate to true (which I think is overkill) to cover an unusual case in smp. There's a scenario in which an address gets unmapped by one cpu and then mapped by another cpu but which hasn't run the shootdown yet so the old translation is still in the TLB. The page table is updated, but code run after the map would still have the old physical address causing problems. IIRC, this happened before with an interaction between the pagecache and storage drivers both of which were doing a lot of alloc/dealloc. I haven't tripped it yet though with this branch, so perhaps it's mitigated in a way I haven't found yet.

Reply:
I'm just circling back to this after the fault digression. I think this code would invalidate in the case you describe, though maybe I'm not understanding the scenario. The invalidate flag is just to only invalidate for actual page (or block / large page) addresses and not for intermediate tables; it's only set to false when installing a new table page.

Reply:
Yes, it looks like you're right. I don't see a way that happens now, I think I was mistaken before.  Thanks!